### PR TITLE
TypeScript Code generation

### PR DESCRIPTION
**Pull request by rricard**
This is a work in progress targeting TS type annotation generation. For now, only basic structures will be generated.

### CODE REVIEW

**Comment by stubailo** [comment]
I think it will be easier to see what it generates with some tests!

**Comment by rricard** [comment]
@stubailo I intend to write a lot of them, I'm just starting on this one. I just opened the PR in case you want to provide some early feedback.

**Comment by stubailo** [comment]
Yeah, my main interest would be to see some example TS code generated from this approach, and one way would have been to look at some tests! But I can wait :]

**Comment by stubailo** [comment]
Yeah, I think for the iOS generation we took a different approach which was to design the generated result first, then implement the generation. But either way is fine!

**Comment by rricard** [comment]
@stubailo I'm trying to restart with the implementation structure for swift actually. What helps is that I clearly know what I want to get but I should formalize it indeed. I'll try to do that before continuing.

**Comment by rricard** [comment]
@stubailo found a basic query passing with the generator I currently have 😉 : https://github.com/apollostack/apollo-codegen/blob/a9ef9c1d1b351701426209edb17d215848194cb8/test/typescript/codeGeneration.js#L62:L79

**Comment by rricard** [comment]
and ... actually writing it in a test makes me realize it's not valid TS 😉. Thank you very much for the advice! Will continue on TDD now, I usually do that but I don't know why, I did it freestyle this time!

**Comment by martijnwalraven** [comment]
@rricard: Seems like you're on a roll! Let me know if you have any questions about the codebase, or if anything needs to change about the compiler IR output.

**Comment by rricard** [comment]
here is what I'll have for today: https://github.com/apollostack/apollo-codegen/blob/8bda617e7b69ffecfc2cf716cba7247260f7a958/test/typescript/codeGeneration.js

**Comment by rricard** [comment]
@martijnwalraven seems good for now, I'll definitely need to refactor at some point, the swift code structure was a good start but TS doesn't need that much stuff (I'm just generating type annotations) so I'll try to see how to make things lighter ... But, yea, the generator is definitely a good idea, especially for indents and blocks!

**Comment by martijnwalraven** [comment]
Good to hear! I'll probably try to split up the Swift `codeGeneration.js` at some point as well, because it is indeed getting a little too heavy.

**Comment by rricard** [comment]
Fearing to do the same in TS...

**Comment by rricard** [comment]
@stubailo would love some feedback on what I'm generating (see tests 🎉 !). Once you feel it's what we would expect from this generator, I'll refactor it!

**Comment by rricard** [comment]
I need to test that against the kind of complex queries I have in my app and see if that doesn't break (we have TONS of fragments!)

File: test/typescript/codeGeneration.js [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+          hero {
+            name
+          }
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
```
Comment (stubailo) [inline test/typescript/codeGeneration.js]: I think it would be best to write tests so that they don't all have to hardcode this boilerplate at the top.

**Comment by rricard** [comment]
Ok, just tested excepted small edge cases, I think this works pretty well!

Here is what I see that is left to cover:

- [ ] Nullable arrays of nullable elements (for now, doing something ugly such as: `Item | null[] | null`, should do: `(Item | null)[] | null`)
- [ ] Input object fields (check if they work correctly)

File: test/typescript/codeGeneration.js [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+        // The episodes in the Star Wars trilogy
+        export type Episode =
```
Comment (stubailo) [inline test/typescript/codeGeneration.js]: If we compile multiple queries with this enum, do we generate the type multiple times, or just once? Sorry, I'm not really in depth on the codegen, although I'm trying to learn more about it!

File: test/typescript/codeGeneration.js [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        export interface HeroAndFriendsNamesQueryVariables {
+          episode: Episode | null;
+        }
+
+        export interface HeroAndFriendsNamesQuery {
+          hero: {
+            name: string,
+            friends: Array< {
+              name: string,
+            } > | null,
```
Comment (stubailo) [inline test/typescript/codeGeneration.js]: Personally I'd prefer `{ ... }[]` over `Array<{ ... }>`, just because of the JSX thing.

File: test/typescript/codeGeneration.js [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        export interface HeroAndFriendsNamesQuery {
+          hero: HeroFriendsFragment & {
```
Comment (stubailo) [inline test/typescript/codeGeneration.js]: Wow, TypeScript is great!

File: test/typescript/codeGeneration.js [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        export interface HeroAndDetailsQuery {
+          hero: HeroDetailsFragment & {
+            name: string,
+          } | null;
+        }
+
+        export interface HeroDetailsFragment {
```
Comment (stubailo) [inline test/typescript/codeGeneration.js]: Would it be more accurate to do something like:

```ts
{ primaryFunction: string | null } | { height: number | null }
```

So that it encodes that it's exactly one of the two combinations?

**Comment by stubailo** [comment]
@jbaxleyiii take a look at this as well, it could be a good base for your Flow Type generator!

**Comment by rricard** [comment]
@stubailo & @jbaxleyiii I actually need a flow generator as well and I plan to make this one run either in flow or ts mode once I hit the next PR 😉 . Since both syntaxes are similar, I think we could exploit the same thing to generate them!

File: test/typescript/codeGeneration.js (hunk 2 above)
Comment (rricard) [inline test/typescript/codeGeneration.js]: This concern is actually handled "generally" in the common parts of the codegen system. @martijnwalraven may be more able to tell you how he did that but in my codebase for instance I'm reusing the same enum type at multiple places and it's generated on top only once!

File: test/typescript/codeGeneration.js (hunk 3 above)
Comment (rricard) [inline test/typescript/codeGeneration.js]: @stubailo actually cheating here to get something closer to what I would do in flow ...

Comment (stubailo) [inline test/typescript/codeGeneration.js]: That's a fine justification!

File: test/typescript/codeGeneration.js (hunk 2 above)
Comment (stubailo) [inline test/typescript/codeGeneration.js]: Sounds great to me! I was suspecting Martijn's code already did the right thing ;]

File: test/typescript/codeGeneration.js [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+          }
+        }
+      `);
+
+      const source = generateSource(context);
+
+      expect(source).to.include(stripIndent`
+        //  This file was automatically generated and should not be edited.
+
+        // The episodes in the Star Wars trilogy
```
Comment (rricard) [inline test/typescript/codeGeneration.js]: @stubailo interesting, I would love to get doc comments as well ... Do you have any doc on that subject so I can add it correctly ?

File: test/typescript/codeGeneration.js (hunk 2 above)
Comment (rricard) [inline test/typescript/codeGeneration.js]: @martijnwalraven I'm amazed on the amount of stuff I don't have to do! Thank you, that's some quality code you wrote!

File: test/typescript/codeGeneration.js (hunk 1 above)
Comment (rricard) [inline test/typescript/codeGeneration.js]: agreed, will try to see that during the refactor!

File: test/typescript/codeGeneration.js (hunk 6 above)
Comment (stubailo) [inline test/typescript/codeGeneration.js]: OK, I just tested it out. It looks like the comment has to be of the form `/** content */` for VSCode to recognize it and display as a tooltip.

Comment (stubailo) [inline test/typescript/codeGeneration.js]: It even supports markdown!!

```ts
/**
 * Generate an executable schema
 *
 * - This
 * - is
 * - A list
 */
function makeExecutableSchema({
  typeDefs,
  resolvers = {},
  connectors,
  logger,
  allowUndefinedInResolve = true,
  resolverValidationOptions = {},
}: IExecutableSchemaDefinition) {
```

Comment (rricard) [inline test/typescript/codeGeneration.js]: gotcha! will do that as much as I can, I think I'll do that on fields as well. This is super useful!

Comment (rricard) [inline test/typescript/codeGeneration.js]: @stubailo GraphQL descriptions are actual markdown: this will be fun! I just need to create an helper to create multiline doc comments!

File: test/typescript/codeGeneration.js (hunk 5 above)
Comment (stubailo) [inline test/typescript/codeGeneration.js]: Yeah, I don't think it's critical. Although I think Martijn is doing something like this in the iOS code, so maybe there is an easy way?

Comment (rricard) [inline test/typescript/codeGeneration.js]: I'll keep that in mind while refactoring, I still need to get more familiar with all of this!

File: test/typescript/codeGeneration.js (hunk 6 above)
Comment (rricard) [inline test/typescript/codeGeneration.js]: Urrgh, doesn't seem to work with attributes... Feels like only functions/methods will work...

Comment (stubailo) [inline test/typescript/codeGeneration.js]: Dang. Well, we can iterate on it later.

**Comment by pspeter3** [comment]
One question I have is whether or not `null` will be explicitly sent back from the server. If not, the conditional property operator may be better.

**Comment by rricard** [comment]
@pspeter3 usually I see that the server sends a null back, that's how the graphql execution engine works in the spec at least

**Comment by rricard** [comment]
@pspeter3 btw, if you want to work on tests, you should fork the repo and derive your own branch from this one (basically, you will end up doing a PR into this PR :wink:), that way, I'll be able to see through the diff more easily and guide you more easily!

**Comment by rricard** [comment]
Here is a list I rapidly put together of what's left to do:

- more tests on working features (take examples from the swift tests)
- tests on non-working features (you will have to figure out the syntax we want - so don't hesitate to ask me, either here or on slack - my username is @rricard)
- I may work on making those tests pass but if you find yourself a lot of time to contribute, you can do it as well!

Any question: ping me here or on slack. Just keep in mind that I'm in Europe so, well, I may sleep while you are working!

**Comment by pspeter3** [comment]
I have my own fork already and was planning to do the PR of a PR. Also one question, why decided to do type aliases over interfaces? Was it to intersect fragments? My belief was that type aliases don't show up in compilation errors so we may have a better developer experience if we use interfaces.

**Comment by rricard** [comment]
@pspeter3 most of my decisions are based on doing something similar to flow (I am using that primarily, but I wanted to do TS first to see later how much I could centralize under the same syntax actually). So, in the test, do what you feel is the more logical for TS and I'll figure out how to make that common with flow and TS later. My first objective will be to make your tests pass. So, please write some idiomatic TS, and I'll work around that!

**Comment by rricard** [comment]
That's actually great to have you give your input on this. I think we could end up with a better implementation in the end!

**Comment by dotansimha** [comment]
Tried some use cases for the generator, it looks good, but there are some issues I noticed:

- When using custom inline fragment on an object that of type List (defined as `[MyObject]`) in the scheme definition, the generated type does not handle the fragment as array. 
This is a unit test that fails, to demonstrate the issue:
```
    it(`should generate correct list with custom fragment`, function() {
      const context = this.compileFromSource(`
        fragment Friend on Character {
          name
        }
        
        query HeroAndFriendsNames($episode: Episode) {
          hero(episode: $episode) {
            name
            friends {
              ...Friend
            }
          }
        }
      `);

      const source = generateSource(context);

      expect(source).to.include(stripIndent`
        //  This file was automatically generated and should not be edited.
        
        // The episodes in the Star Wars trilogy
        export type Episode =
          "NEWHOPE" | // Star Wars Episode IV: A New Hope, released in 1977.
          "EMPIRE" | // Star Wars Episode V: The Empire Strikes Back, released in 1980.
          "JEDI"; // Star Wars Episode VI: Return of the Jedi, released in 1983.
        
        
        export interface HeroAndFriendsNamesQueryVariables {
          episode: Episode | null;
        }
        
        export interface HeroAndFriendsNamesQuery {
          hero: {
            name: string,
            friends: Array<FriendFragment>,
          } | null;
        }
        
        export interface FriendFragment {
          name: string;
        }
      `);
    });
```

**Comment by rricard** [comment]
@dotansimha that's exactly the kind of feedback I needed! thanks! I'm back working on this this week! 😉

**Comment by dotansimha** [comment]
Found a solution, need to replace the following line:
```ts
https://github.com/apollostack/apollo-codegen/blob/e32dd8983ec8733f76c7843ea66c26f8ba1ecc69/src/typescript/language.js#L38
```

to:
```ts
    generator.printOnNewline(`${propertyName}: ${isArray ? 'Array<' : ''}${fragmentSpreads.map(n => `${n}Fragment`).join(' & ')}${isArray ? '>' : ''}`);
```

Will create a PR into this PR in a few minutes.

@rricard , any chance to add a TODO list or a list of known issues to this PR? :)

**Comment by rricard** [comment]
@dotansimha Nice! Please add a test for your case if you write a PR. I don't have any TODO right now but you are right it would be useful, I'll start by doing that this afternoon!

**Comment by rricard** [comment]
Okay nice, let's finish this PR then! Just going to eat something and I'm back at work!

**Comment by dotansimha** [comment]
Also noticed that the generator changed the name of the fields (and it should not do it in any case - it's fine for the name of the operation or interface, but not for the field name).
For example, a field named `full_name` will generate a type with a field called `fullName`.

**Comment by dotansimha** [comment]
Also, @rricard , the concept of generating one big interface is a mistake. 

The actual usage will be something like:
```ts
    this.subscriptionSub = this.apollo.subscribe({
      query: subscriptionQuery,
      variables: {repoFullName: repoName},
    }).subscribe({
      next: (data: OnCommentAddedSubscription) => {
        const newComment: Comment = data.commentAdded;
        // ... more code ...
      }
    });
```

`OnCommentAddedSubscription` is generated, and looks like that:
```ts
export interface OnCommentAddedSubscription {
  commentAdded: {
    id: number,
    postedBy: {
      login: string,
      html_url: string,
    },
    createdAt: number,
    content: string,
  } | null;
}
```

The problem is that `newComment` it still remains as `any` because we can't use the inner type of `OnCommentAddedSubscription` as interface, the actual generated should looks like that:

```ts
export interface OnCommentAddedSubscriptionCommentAdded {
{
    id: number,
    postedBy: {
      login: string,
      html_url: string,
    },
    createdAt: number,
    content: string,
  }
}

export interface OnCommentAddedSubscription {
  commentAdded: OnCommentAddedSubscriptionCommentAdded | null;
}
```

So when you want to use an inner type in your code - it's an easy usage. 

The level of generated code can depend on the implementation, but the first level must get it's own interface so you will be able to extract inner types from the GraphQL response.

**Comment by rricard** [comment]
@dotansimha I prefer the module notation. For now, I don't think we should change this. I want to ship that as fast as possible and actually derivate this implementation for Flow ASAP (which in its current state won't take much time).

In your example, I don't see the issue though, just let TS infer your type right? Or did I spend too much time with Flow and its inference engine that I think it's a completely normal thing?

Anyway, I see the point of what you want to do, maybe just let's do that on the next iteration for TS support!

**Comment by rricard** [comment]
I think a "first version" is quite ready now, thanks to you (for handling those edge cases that I missed), I'll let @stubailo or @martijnwalraven give there feedback and then, let's get this merged.

Thanks for your help, ping me on slack to discuss what's next on this (same as on github: @rricard)!

**Comment by rricard** [comment]
@dotansimha ping me on slack if you see anything else that seems important to you!

I'll work on refactoring that PR a bit once we agree it covers a "minimal" TS support...

**Comment by dotansimha** [comment]
TS with WebStorm does not auto complete if you don't declare the type, so when using an inner property of an object - you must specify it's type, otherwise it treated as `any` and the auto-complete for fields does not work :(

**Comment by rricard** [comment]
Humm, well, I'll see what I can do during the refactor then, do you see anything else we missed before I start?

**Comment by pspeter3** [comment]
In general, we should prefer `namespace` over `module`.

**Comment by rricard** [comment]
Okay, I added input object types. I will followup tomorrow with a proper refactor!

**Comment by rricard** [comment]
@martijnwalraven & @stubailo maybe we should merge that now, the feature is properly isolated and this would allow the following:

- other community contributors will be able to propose PRs against master
- I'll be able to start the flow implementation (and refactor things while doing it)
- I think that proposing the feature to try as early as possible would help us spot bugs or things wanted by the users more easily

**Comment by rricard** [comment]
(in fact starting with flow ASAP may be beneficial since it would push me faster towards refactor)

**Comment by martijnwalraven** [comment]
@rricard: Happy to merge this, looks great!

**Comment by stubailo** [comment]
Wow, excited to see so much progress today on this!

**Comment by rricard** [comment]
@stubailo you can thank @dotansimha that spotted things I couldn't think of! Hope he'll stay around to help us get a better TS support!

**Comment by martijnwalraven** [comment]
@rricard: I think it would be great if you could write an article about this for the Apollo blog!

**Comment by rricard** [comment]
@martijnwalraven Good idea! I need to setup a sample project first though to be able to try out things correctly!

**Comment by csillag** [comment]
+1

**Comment by rricard** [comment]
@martijnwalraven can we have a release on npm with typescript support so I can use it in a sample repo? Thanks!

**Comment by martijnwalraven** [comment]
@rricard: Just published apollo-codegen 0.9.5 on npm

**Comment by rricard** [comment]
@martijnwalraven :+1: eating something and then I'm back on it!
